check.sh

```diff
@@ -11,44 +11,6 @@ if [ -z "${GITHUB_STEP_SUMMARY+x}" ]; then
     ON_GITHUB_CI=false
 fi
 
-# Test if the generated code is still up to date
-echo "::group::Generate Exports"
-python ./src/trio/_tools/gen_exports.py --test \
-    || EXIT_STATUS=$?
-echo "::endgroup::"
```

Comment:
I probably should do this in another PR but I think we should move gen exports out of pre-commit and back here. Unfortunately, as things currently are, if you modify a file that the pre-commit hook says it uses then commit with a git client (ie not just `git commit -m "..."` in the venv you have for trio), gen_exports doesn't get the dependencies it needs.

So far I've been getting that error then going to my terminal to run git there, but that's a bit much.

Reply:
Pre-commit with a external git client doesn't install required dependencies?

Reply:
Nope, because we don't specify them. We don't specify them because pre-commit doesn't let you install from a requirements file :(

Reply:
I specified them in the latest commit: https://github.com/python-trio/trio/blob/17d3d051fe7486a8f3731852a14438b87046cf60/.pre-commit-config.yaml#L46

but that doesn't let us pin the versions (in a non-awful way), so yeah if we don't want that then it has to go outside of pre-commit

Reply:
> https://github.com/python-trio/trio/blob/17d3d051fe7486a8f3731852a14438b87046cf60/.pre-commit-config.yaml#L46

Keep in mind that the pre-commit cache is virtually immortal. You can clean it locally, but not in the `pre-commit.ci` service. For that, changing the deps list or the hook version is required. This may be problematic sometimes.

So I've found that the best thing to do here is to keep the version pins in the list and bump them periodically. Just so that the cache doesn't grow too old.

Reply:
I think that means we need to keep it in `check.sh` and skip it in ci for now. I'll keep the changes to the pre-commit hook to alleviate @A5rocks's problem, and maybe use `tox` or something later on.

Reply:
Tox sounds wonderful. I've been meaning to suggest it. I've actually been working on some reusable workflows for tox-based projects + experimented with some cool GHA integrations.
When you get to it, I may offer some help / experimentation in simplifying things here.

Reply:
Yeah I'm personally a big fan of tox. #2790 discussed hatchling too but we ended up bikeshedding too much to actually do anything in the end.

Reply:
I've yet to use nox, but would def be open to.

Reply:
I've contributed to projects with nox, and they're slightly more inconvenient. I'd also want to run nox python files under coverage.

one thing to bear in mind with tox is that constraints are a bit of a pain, tox-uv and a lock file seems the way to go:

https://github.com/tox-dev/tox-uv?tab=readme-ov-file#uvlock-support

Reply:
I've been taking DIY lock files to extremes with tox for a while. Haven't yet tried uv for that, really. I know that they were implementing the poerty-style format, not real lock files. I'm looking forward to seeing Brett's PEP accepted, honestly.

I also have a lot more experience with tox and haven't contributed to nox. Although, I have contributed to projects using nox. There's limitations for parallelism, for example. Some scripting would be more convenient in Python. However, tox 4 has a `toxfile.py` that should be able to cover a lot of that. I'm yet to play with it properly.